contracts/adapters/OptimisticAuctionRebalanceExtensionV1.sol

```diff
@@ -0,0 +1,406 @@
+    * when true it is open.
+    *
+    * @param _isOpen           Bool indicating whether the extension is open for proposing rebalances.
+    */
+   function updateIsOpen(bool _isOpen) external onlyOperator {
+       isOpen = _isOpen;
+    }
+
+
+
+    /**
+    * @dev OPERATOR ONLY: sets product settings for a given set token
+    * @param _optimisticParams OptimisticRebalanceParams struct containing optimistic rebalance parameters.
+    * @param _rulesHash bytes32 containing the ipfs hash rules for the product.
+    */
+    function setProductSettings(
+        OptimisticRebalanceParams memory _optimisticParams,
+        bytes32 _rulesHash
+    )
+        external
+        onlyOperator
+    {
+        productSettings = ProductSettings({
+            optimisticParams: _optimisticParams,
+            rulesHash: _rulesHash
+        });
+        emit ProductSettingsUpdated(setToken, setToken.manager(), _optimisticParams, _rulesHash);
+    }
+
+     /**
+     * @param _quoteAsset                   ERC20 token used as the quote asset in auctions.
+     * @param _oldComponents                Addresses of existing components in the SetToken.
+     * @param _newComponents                Addresses of new components to be added.
+     * @param _newComponentsAuctionParams   AuctionExecutionParams for new components, indexed corresponding to _newComponents.
+     * @param _oldComponentsAuctionParams   AuctionExecutionParams for existing components, indexed corresponding to
+     *                                      the current component positions. Set to 0 for components being removed.
+     * @param _rebalanceDuration            Duration of the rebalance in seconds.
+     * @param _positionMultiplier           Position multiplier at the time target units were calculated.
+    */
+    function proposeRebalance(
+        IERC20 _quoteAsset,
+        address[] memory _oldComponents,
+        address[] memory _newComponents,
+        AuctionExecutionParams[] memory _newComponentsAuctionParams,
+        AuctionExecutionParams[] memory _oldComponentsAuctionParams,
+        uint256 _rebalanceDuration,
+        uint256 _positionMultiplier
+    )
+        external
+        onlyAllowedAssets(_newComponents)
+        onlyIfOpen()
+    {
+        bytes32 proposalHash = keccak256(abi.encode(
+            setToken,
+            _quoteAsset,
+            _oldComponents,
+            _newComponents,
+            _newComponentsAuctionParams,
+            _oldComponentsAuctionParams,
+            false, // We don't allow locking the set token in this version
+            _rebalanceDuration,
+            _positionMultiplier
+        ));
+        require(assertionIds[proposalHash] == bytes32(0), "Proposal already exists");
+        require(productSettings.rulesHash != bytes32(""), "Rules not set");
+        require(address(productSettings.optimisticParams.optimisticOracleV3) != address(0), "Oracle not set");
```

Comment:
Let's add unit tests that trip these require statements

Reply:
Done: https://github.com/IndexCoop/index-coop-smart-contracts/commit/c8a85859cc8c185b5238cc9928eb7c41efcb7e7f